### PR TITLE
refactor(cache): improve cache and integrate reagent

### PR DESCRIPTION
**Pull request by smothers**
## PR Description

The current homebase cache is not very composable and it's kind of hard to reason about.

The goal of this PR is to rewrite the cache to make it more readily support different view frameworks like (React or Reagent) and different DBs like (Datascript or Datahike).

- [ ] Cache
    - [ ] homebase.cache
        - [x] Entity + Attribute cache
        - [x] Query cache
        - [ ] Ensure that view updates are also transactional and consistent
            - [x] Pass the same DB snapshot to all the change-handlers of a given transaction
            - [ ] Batch all UI updates to occur simultaneously
- [ ] Views
    - [ ] homebase.react
        - [ ] useEntity
        - [ ] useQuery
        - [ ] remove old cache code
    - [x] homebase.reagent
        - [x] (homebase.reagent/connect! db-conn)
        - [x] (homebase.reagent/entity db-conn lookup)
        - [x] reactive homebase.reagent/Entity type
        - [x] (homebase.reagent/q query db-conn & args)
        - [ ] Release on Clojars
- [ ] DBs
    - [x] Datascript
    - [ ] Datahike (probably save this for another PR)
- [ ] Tests
    - [x] Reagent UI tests for all the different reactivity patterns (entity, ref, reverse ref, q, q + pull)
    - [ ] Reagent UI tests to make sure the minimal re-render is happening and the dom is not being thrashed
    - [ ] Datahike + (Reagent / React) tests

### Testing
- [x] added relevant test coverage
- [ ] no tests needed

### Docs
- [x] added relevant docs 
    - preview them at https://homebase.io/docs/homebase-react/{BRANCH_NAME}/overview
- [x] updated relevant sections in the README.md
- [ ] updated relevant docstrings in index.d.ts
- [ ] no docs needed

### Typescript
- [ ] added or edited relevant Typescript type declarations
- [x] no type declaration updates needed

## Merging
For maintainers.

To merge, select "Squash and Merge". Then:
  1. Make sure the top commit message follows [Angular Git Commit Guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines),
  2. Delete all other commit messages in the description, but keep any lines designating [co-authors](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors) so contributors will retain credit for their contributions.

### CODE REVIEW

File: src/homebase/js.cljs [hunk 1]
```diff
@@ -97,22 +98,22 @@
...
-          [:db/add id :homebase.array/order (+ 1 i)]]
-         (if (scalar? v)
-           [[:db/add id :homebase.array/value v]]
-           (let [child-id (or (get v "id")
-                              (get (second (first v)) "id")
-                              (swap! temp-ids-atom dec))]
-             (into [[:db/add id :homebase.array/ref child-id]]
-                   (js->tx-part schema temp-ids-atom (cons [attr child-id true] key-path) v)))))))
-    tx-part)))
+          (map-indexed
```
Comment (rubinovitz) [inline src/homebase/js.cljs:101]: @smothers looks like there is some whitespace here

File: src/homebase/react.cljs [hunk 2]
```diff
@@ -29,13 +28,13 @@
                              (clojure.string/trim))))))))))
 
 (defn debug-msg [return-value & msgs]
-  (when (and (number? hbjs/*debug*) (>= hbjs/*debug* 2))
```
Comment (rubinovitz) [inline src/homebase/react.cljs:32]: @smothers whitespace

File: src/homebase/react.cljs [hunk 3]
```diff
@@ -29,13 +28,13 @@
...
 
 (defn debug-msg [return-value & msgs]
-  (when (and (number? hbjs/*debug*) (>= hbjs/*debug* 2)) 
+  (when (and (number? hbjs/*debug*) (>= hbjs/*debug* 2))
     (apply js/console.log "%c homebase-react " "background: yellow" msgs))
   return-value)
 
 (defn changed? [entities cached-entities track-count?]
   (cond
-    (and track-count?
```
Comment (rubinovitz) [inline src/homebase/react.cljs:38]: @smothers more whitespace.

File: src/homebase/react.cljs [hunk 4]
```diff
@@ -137,10 +136,10 @@
         conn (d/create-conn (if schema
                               (merge (hbjs/js->schema schema) base-schema)
                               base-schema))]
-    (datalog-console/init! {:conn conn})
+    (datalog-console/enable! {:conn conn})
     (when initial-tx (hbjs/transact! conn initial-tx))
     (react/createElement
-     (goog.object/get homebase-context "Provider")
```
Comment (rubinovitz) [inline src/homebase/react.cljs:143]: @smothers whitespace

File: src/homebase/js.cljs (hunk 1 above)
Comment (smothers) [reply src/homebase/js.cljs:101]: yes, the whitespace was off before

**Review by rubinovitz — APPROVED** [review]
LGTM.
I'm most qualified to comment on the docs and they look great. Excited to get this out there!!!

**Pull request merged**